eth-repository/src/main/kotlin/net/consensys/cava/eth/repository/BlockchainRepository.kt

```diff
@@ -0,0 +1,241 @@
+  ) {
+
+  companion object {
+
+    /**
+     * Initializes a blockchain repository with metadata, placing it in key-value stores.
+     *
+     * @return a new blockchain repository made from the metadata passed in parameter.
+     */
+    suspend fun init(
+      blockStore: KeyValueStore,
+      blockHeaderStore: KeyValueStore,
+      chainMetadata: KeyValueStore,
+      blockchainIndex: BlockchainIndex,
+      genesisBlock: Block
+    ): BlockchainRepository {
+      val repo = BlockchainRepository(chainMetadata, blockStore, blockHeaderStore, blockchainIndex)
+      repo.setGenesisBlock(genesisBlock)
+      repo.storeBlock(genesisBlock)
+      return repo
+    }
+  }
+
+  /**
+   * Stores a block into the repository.
+   *
+   * @param block the block to store
+   * @return a handle to the storage operation completion
+   */
+  suspend fun storeBlock(block: Block) {
+    blockStore.put(block.header().hash().toBytes(), block.toBytes())
+    blockHeaderStore.put(block.header().hash().toBytes(), block.header().toBytes())
+    blockchainIndex.indexBlockHeader(block.header())
+    if (isChainHead(block.header())) {
+      setChainHead(block.header())
+    }
+  }
+
+  /**
+   * Stores a block header in the repository.
+   *
+   * @param header the block header to store
+   * @return handle to the storage operation completion
+   */
+  suspend fun storeBlockHeader(header: BlockHeader) {
+    blockHeaderStore.put(header.hash().toBytes(), header.toBytes())
+    if (isChainHead(header)) {
+        setChainHead(header)
+    }
+  }
+
+  private suspend fun isChainHead(header: BlockHeader): Boolean {
+    val headHeader = retrieveChainHeadHeader()
+    return headHeader?.number()?.compareTo(header.number()) == -1
+  }
+
+  /**
+   * Retrieves a block into the repository as its serialized RLP bytes representation.
+   *
+   * @param blockHash the hash of the block stored
+   * @return a future with the bytes if found
+   */
+  suspend fun retrieveBlockBytes(blockHash: Hash): Bytes? {
+    return retrieveBlockBytes(blockHash.toBytes())
+  }
+
+  /**
+   * Retrieves a block into the repository as its serialized RLP bytes representation.
+   *
+   * @param blockHash the hash of the block stored
+   * @return a future with the bytes if found
+   */
+  suspend fun retrieveBlockBytes(blockHash: Bytes): Bytes? {
+    return blockStore.get(blockHash)
+  }
+
+  /**
+   * Retrieves a block into the repository.
+   *
+   * @param blockHash the hash of the block stored
+   * @return a future with the block if found
+   */
+  suspend fun retrieveBlock(blockHash: Hash): Block? {
+    return retrieveBlock(blockHash.toBytes())
+  }
+
+  /**
+   * Retrieves a block into the repository.
+   *
+   * @param blockHash the hash of the block stored
+   * @return a future with the block if found
+   */
+  suspend fun retrieveBlock(blockHash: Bytes): Block? {
+    val bytes: Bytes? = retrieveBlockBytes(blockHash)
+    return if (bytes == null) {
```

Comment:
return `bytes?.let { b -> Block.fromBytes(b) }`

Reply:
Or, perhaps better, `val bytes = retrieveBlockBytes(blockHash) ?: return null` in the line above